### PR TITLE
Add Scylla Operator demo

### LINKED ISSUE #1193: Add Scyla Operator Demo
**Is this a bug report or feature request?**
* Feature Request

**What should the feature do:**
Add an asciicast that we can put into a readme on how to setup the operator and create a ScyllaCluster.

**What is use case behind this feature:**
UX

**Additional Information:**
As a lot of the setup is prone to change the asciicast should be scripted and easily updatable.

### PR DESCRIPTION
**Pull request by tnozicka**
**Description of your changes:**
This PR adds a demo recording on scylla-operator installation and some basic operations.

The full demo is scripted so it should be easy to modify the instruction and recapture the demo if we change the install or want to extend the demo.

Fixes #1193

### CODE REVIEW

**Comment by rzetelskik** [comment]
Should we mention it in README and/or docs?

**Comment by tnozicka** [comment]
> Should we mention it in README and/or docs?

you are correct, but I plan to do it as a next step as we need to fix our deploy instructions first, along site the new docs

**Comment by rzetelskik** [comment]
I also wonder if we could prevent it from sharing the fate of our documentation and just gradually getting out of date and broken... because I know for sure we're not going to run this ourselves other than for presentations.

**Comment by tnozicka** [comment]
there will likely be a GKE job making sure it works, probably periodic - also a reason why this is scripted - so it's testable and easy to update

File: recordings/basic-setup/run.sh [hunk 1]
```diff
@@ -0,0 +1,309 @@
...
+
+run <<< "# Because all the following manifests use default storage class, we'll set scylladb-local-xfs as the default one, for simplicity."
+run <<< "# As there can be only one default storage class, make sure to unset the previous default one, if it existed in your setup."
+run <<< "# This can be also controlled per ScyllaCluster."
+run <<'EOF'
+kubectl patch storageclass scylladb-local-xfs -p '{"metadata": {"annotations":{"storageclass.kubernetes.io/is-default-class":"true"}}}'
+EOF
+sleep-extra 5
+
+run <<< "# Optionally, deploy Scylla Manager"
```
Comment (rzetelskik) [inline recordings/basic-setup/run.sh:178]:
```suggestion
run <<< "# Optionally, deploy Scylla Manager."
```
nit, applies globally (some have it, some don't)

File: recordings/basic-setup/run.sh [hunk 2]
```diff
@@ -0,0 +1,309 @@
...
+EOF
+
+run <<< "# We have now successfully deployed the infrastructure and can start creating ScyllaDB clusters."
+
+run <<< "# Create demo namespace"
+run <<'EOF'
+kubectl create namespace demo --dry-run=client -o=yaml | kubectl apply --server-side -f=-
+EOF
+
+run <<< "# Optionally, create a config for ScyllaDB cluster"
```
Comment (rzetelskik) [inline recordings/basic-setup/run.sh:194]:
```suggestion
run <<< "# Optionally, create a config for ScyllaDB clusters."
```
nit, for consistency

File: recordings/basic-setup/run.sh [hunk 3]
```diff
@@ -0,0 +1,309 @@
...
+
+run <<< "# Create a ScyllaDB cluster (ScyllaCluster)"
+run <<'EOF'
+yq -e eval --colors '.' ./examples/scylladb/scylla.scyllacluster.yaml | head -n 25
+EOF
+sleep-extra 5
+run <<'EOF'
+kubectl -n=demo apply --server-side -f=./examples/scylladb/scylla.scyllacluster.yaml
+EOF
+run <<< "# Wait for the ScyllaCluster to rollout"
```
Comment (rzetelskik) [inline recordings/basic-setup/run.sh:211]:
```suggestion
run <<< "# Wait for the ScyllaCluster to roll out."
```

File: recordings/basic-setup/run.sh [hunk 4]
```diff
@@ -0,0 +1,309 @@
...
+run <<'EOF'
+monitoring_manifest="$( mktemp )"
+EOF
+run <<'EOF'
+yq -e eval '. | .spec.endpointsSelector.matchLabels["scylla/cluster"] = "scylla" | .spec.components.grafana.exposeOptions.webInterface.ingress.dnsDomains = ["example-grafana.demo.svc.cluster.local"]' ./examples/monitoring/v1alpha1/scylladbmonitoring.yaml | tee "${monitoring_manifest}" | yq -e eval --colors '.' | head -n 25
+EOF
+run <<'EOF'
+kubectl -n=demo apply --server-side -f="${monitoring_manifest}"
+EOF
+run <<< "# Wait for the ScyllaDBMonitoring to rollout"
```
Comment (rzetelskik) [inline recordings/basic-setup/run.sh:246]: ditto

File: recordings/basic-setup/run.sh [hunk 5]
```diff
@@ -0,0 +1,309 @@
...
+run <<'EOF'
+kubectl -n=demo wait --for='condition=Available' --timeout=10m scylladbmonitoring.scylla.scylladb.com/example
+EOF
+run <<'EOF'
+kubectl -n=demo rollout status --timeout=10m deployment.apps/example-grafana
+EOF
+run <<'EOF'
+kubectl -n=demo get svc/example-grafana
+EOF
+run <<< "# Verify we can login to Grafana"
```
Comment (rzetelskik) [inline recordings/basic-setup/run.sh:262]:
```suggestion
run <<< "# Verify we can log in to Grafana."
```

File: recordings/basic-setup/run.sh [hunk 6]
```diff
@@ -0,0 +1,309 @@
...
+curl --fail --retry-all-errors 10 --max-time 3 --retry-max-time 5 --cacert <( echo "${grafana_serving_ca_cert}" ) --resolve 'example-grafana.demo.svc.cluster.local:5000:127.0.0.1' -IL 'https://example-grafana.demo.svc.cluster.local:5000/' --user "${grafana_user}:${grafana_password}"
+EOF
+sleep-extra 3
+
+run <<< "# Scale the ScyllaDB cluster from 1 to 3 nodes"
+run <<'EOF'
+yq -e eval '.spec.datacenter.racks[0].members = 3' ./examples/scylladb/scylla.scyllacluster.yaml | kubectl -n=demo apply --server-side -f -
+EOF
+
+run <<< "# Wait for the ScyllaCluster to rollout"
```
Comment (rzetelskik) [inline recordings/basic-setup/run.sh:286]: ditto

File: recordings/basic-setup/run.sh [hunk 7]
```diff
@@ -0,0 +1,309 @@
...
+  sleep "${CUSTOM_ECHO_SLEEP_SEC}"
+  wait "${min_sleep_pid}"
+  echo
+}
+
+function run {
+  cmd=$( cat )
+  user-echo "${cmd}"
+  eval "${cmd}"
+  run_last_pid=$!
```
Comment (rzetelskik) [inline recordings/basic-setup/run.sh:45]: not used?

File: recordings/basic-setup/run.sh [hunk 8]
```diff
@@ -0,0 +1,309 @@
...
+  sleep "${CUSTOM_COMMAND_SLEEP_SEC}"
+}
+
+function run-watch {
```
Comment (rzetelskik) [inline recordings/basic-setup/run.sh:49]: not used?

**Review by rzetelskik — COMMENTED** [review]
Should we have a very basic readme (with requirements etc.)?

File: recordings/basic-setup/run.sh (hunk 2 above)
Comment (tnozicka) [reply recordings/basic-setup/run.sh:194]: there is just 1 ScyllaDB cluster

File: recordings/basic-setup/run.sh [hunk 9]
```diff
@@ -0,0 +1,309 @@
...
+git clone --depth=1 --branch=master https://github.com/scylladb/scylla-operator.git
+EOF
+  run <<'EOF'
+cd scylla-operator
+EOF
+else
+  cd "${SCYLLA_OPERATOR_REPO}"
+fi
+
+run <<< "# Before we start the deployment, make sure your Kubernetes nodes for running ScyllaDB are labeled the same way to match the manifests."
```
[the remark this answers is not in the document]
Comment (tnozicka) [reply recordings/basic-setup/run.sh:86]: > dedicated to

they don't have to be dedicated + might get confused with taints

> accordingly

seems more vague to me, so I'll keep this one

File: recordings/basic-setup/run.sh [hunk 10]
```diff
@@ -0,0 +1,309 @@
...
+if [[ "${scylla_nodes}" == "" ]]; then
+  echo "Labeling all nodes to be used for ScyllaDB."
+  kubectl label nodes scylla.scylladb.com/node-type=scylla --all
+else
+  printf "Using nodes %q for ScyllaDB." "${scylla_nodes}"
+fi
+EOF
+sleep-extra 3
+
+run <<< "# First we'll install our dependencies. If you already have them installed in your cluster, you can skip this step."
```
[the remark this answers is not in the document]
Comment (tnozicka) [reply recordings/basic-setup/run.sh]: done

File: recordings/basic-setup/run.sh [hunk 11]
```diff
@@ -0,0 +1,309 @@
...
+crd/scyllaclusters.scylla.scylladb.com \
+crd/nodeconfigs.scylla.scylladb.com \
+crd/scyllaoperatorconfigs.scylla.scylladb.com \
+crd/scylladbmonitorings.scylla.scylladb.com
+EOF
+run <<'EOF'
+kubectl -n=scylla-operator rollout status --timeout=10m deployment.apps/{scylla-operator,webhook-server}
+EOF
+
+run <<< "# To run ScyllaClusters on local disk (NVMes in production) we need to setup our storage."
```
[the remark this answers is not in the document]
Comment (tnozicka) [reply recordings/basic-setup/run.sh]: done

File: recordings/basic-setup/run.sh [hunk 12]
```diff
@@ -0,0 +1,309 @@
...
+
+run <<< "# We need to setup our local disks into a XFS directory, that will be used by ScyllaDB local CSI driver to provision PVCs."
```
[the remark this answers is not in the document]
Comment (tnozicka) [reply recordings/basic-setup/run.sh]: mostly applied

File: recordings/basic-setup/run.sh [hunk 13]
```diff
@@ -0,0 +1,309 @@
...
+run <<< "# We will also enable node tuning."
+
+run <<< "# Create a NodeConfig to setup RAID, filesystem and tuning."
```
[the remark this answers is not in the document]
Comment (tnozicka) [reply recordings/basic-setup/run.sh]: done

File: recordings/basic-setup/run.sh [hunk 14]
```diff
@@ -0,0 +1,309 @@
...
+EOF
+
+run <<< "# Wait for the NodeConfig to apply changes to the Kubernetes nodes."
+run <<'EOF'
+kubectl wait --for='condition=Reconciled' --timeout=10m nodeconfigs.scylla.scylladb.com/cluster
+EOF
+# TODO: Extend waiting for the NodeConfig when we fix https://github.com/scylladb/scylla-operator/issues/1557
+
+run <<< "# Install ScyllaDB local CSI driver that will provision storage for PersistentVolumeClaims."
+run <<< "# The ScyllaDB local CSI driver is configured to use folder /mnt/persistent-volumes that was created by the NodeConfig."
```
[the remark this answers is not in the document]
Comment (tnozicka) [reply recordings/basic-setup/run.sh]: done

File: recordings/basic-setup/run.sh [hunk 15]
```diff
@@ -0,0 +1,309 @@
...
+run <<'EOF'
+kubectl -n local-csi-driver apply --server-side -f=https://raw.githubusercontent.com/scylladb/k8s-local-volume-provisioner/v0.2.0-rc.0/{deploy/kubernetes/local-csi-driver/{00_namespace,10_csidriver,10_driver_serviceaccount,10_provisioner_clusterrole,20_provisioner_clusterrolebinding},example/storageclass_xfs}.yaml
+kubectl create -f=https://raw.githubusercontent.com/scylladb/k8s-local-volume-provisioner/v0.2.0-rc.0/deploy/kubernetes/local-csi-driver/50_daemonset.yaml --dry-run=client -o yaml | yq -e '.spec.template.spec.nodeSelector += {"scylla.scylladb.com/node-type": "scylla"}' | kubectl -n=local-csi-driver apply --server-side -f -
+kubectl -n=scylla-operator rollout status --timeout=10m deployment.apps/{scylla-operator,webhook-server}
+EOF
+run <<'EOF'
+kubectl -n=local-csi-driver rollout status --timeout=10m daemonset.apps/local-csi-driver
+EOF
+
+run <<< "# Because all the following manifests use default storage class, we'll set scylladb-local-xfs as the default one, for simplicity."
```
[the remark this answers is not in the document]
Comment (tnozicka) [reply recordings/basic-setup/run.sh]: done

File: recordings/basic-setup/run.sh [hunk 16]
```diff
@@ -0,0 +1,309 @@
...
+run <<< "# As there can be only one default storage class, make sure to unset the previous default one, if it existed in your setup."
```
[the remark this answers is not in the document]
Comment (tnozicka) [reply recordings/basic-setup/run.sh]: done

File: recordings/basic-setup/run.sh (hunk 1 above)
Comment (tnozicka) [reply recordings/basic-setup/run.sh]: done

File: recordings/basic-setup/run.sh (hunk 3 above)
Comment (tnozicka) [reply recordings/basic-setup/run.sh]: done

File: recordings/basic-setup/run.sh (hunk 5 above)
Comment (tnozicka) [reply recordings/basic-setup/run.sh]: done

File: recordings/basic-setup/run.sh (hunk 7 above)
Comment (tnozicka) [reply recordings/basic-setup/run.sh]: removed

File: recordings/basic-setup/run.sh [hunk 17]
```diff
@@ -0,0 +1,292 @@
...
+run <<< "# Wait for the NodeConfig to apply changes to the Kubernetes nodes."
+run <<'EOF'
+kubectl wait --for='condition=Reconciled' --timeout=10m nodeconfigs.scylla.scylladb.com/cluster
+EOF
+# TODO: Extend waiting for the NodeConfig when we fix https://github.com/scylladb/scylla-operator/issues/1557
+
+run <<< "# Install the ScyllaDB local CSI driver that will provision storage for PersistentVolumeClaims."
+run <<< "# The ScyllaDB local CSI driver is configured to use '/mnt/persistent-volumes' directory created by the NodeConfig."
+run <<'EOF'
+kubectl -n local-csi-driver apply --server-side -f=https://raw.githubusercontent.com/scylladb/k8s-local-volume-provisioner/v0.2.0-rc.0/{deploy/kubernetes/local-csi-driver/{00_namespace,10_csidriver,10_driver_serviceaccount,10_provisioner_clusterrole,20_provisioner_clusterrolebinding},example/storageclass_xfs}.yaml
```
Comment (rzetelskik) [inline recordings/basic-setup/run.sh:145]: should we switch to v0.2.0?

Comment (tnozicka) [reply recordings/basic-setup/run.sh:145]: I can do that, but I'll wait for confirmation that everything else is ok, so I avoid rerecording it again

File: recordings/basic-setup/run.sh (hunk 12 above)
[the remark this answers is not in the document]
Comment (rzetelskik) [reply recordings/basic-setup/run.sh]: ScyllaDB Style Guide discourages from using the passive voice, hence my suggestion. Not important to me personally though.

**Comment by rzetelskik** [comment]
lgtm
/assign zimnx

**Comment by zimnx** [comment]
Great demo, thanks!
/lgtm

/hold until k8s local volume provisioner is updated to final version

**Comment by tnozicka** [comment]
I've re-recorded the demo with provisioner v0.2.0 and added `.gitattributes` to mark it as binary file (no diff or line merge)

**Comment by rzetelskik** [comment]
/lgtm
/hold cancel

thanks!

**Pull request merged**